rasa_nlu/persistor.py

```diff
@@ -1,123 +1,177 @@
-from __future__ import unicode_literals
-from __future__ import print_function
-from __future__ import division
 from __future__ import absolute_import
-from builtins import object
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import io
 import os
 import shutil
 import tarfile
-import io
 
 import boto3
 import botocore
-from typing import Optional
-from typing import Text
+from builtins import object
 from rasa_nlu.config import RasaNLUConfig
+from typing import Optional, Tuple
+from typing import Text
 
 
 def get_persistor(config):
     # type: (RasaNLUConfig) -> Optional[Persistor]
-    """Returns an instance of the requested persistor. Currently, `aws` and `gcs` are supported"""
+    """Returns an instance of the requested persistor.
+
+    Currently, `aws` and `gcs` are supported"""
 
     if 'storage' not in config:
-        raise KeyError("No persistent storage specified. Supported values are {}".format(", ".join(['aws', 'gcs'])))
+        raise KeyError("No persistent storage specified. Supported values "
+                       "are {}".format(", ".join(['aws', 'gcs'])))
 
     if config['storage'] == 'aws':
-        return AWSPersistor(config['path'], config['aws_region'], config['bucket_name'], config['aws_endpoint_url'])
+        return AWSPersistor(config['aws_region'], config['bucket_name'],
+                            config['aws_endpoint_url'])
     elif config['storage'] == 'gcs':
-        return GCSPersistor(config['path'], config['bucket_name'])
+        return GCSPersistor(config['bucket_name'])
     else:
         return None
 
 
 class Persistor(object):
     """Store models in cloud and fetch them when needed"""
 
-    def save_tar(self, target_dir):
+    def persist(self, mode_directory, model_name, project):
         # type: (Text) -> None
         """Uploads a model persisted in the `target_dir` to cloud storage."""
-        raise NotImplementedError("")
 
-    def fetch_and_extract(self, filename):
+        if not os.path.isdir(mode_directory):
+            raise ValueError("Target directory '{}' not "
+                             "found.".format(mode_directory))
+
+        file_key, tar_path = self._compress(mode_directory, model_name, project)
+        self._persist_tar(file_key, tar_path)
+
+    def retrieve(self, model_name, project, target_path):
         # type: (Text) -> None
-        """Downloads a model that has previously been persisted to cloud storage."""
+        """Downloads a model that has been persisted to cloud storage."""
+
+        tar_name = self._tar_name(model_name, project)
```

Comment:
I think this doesn't handle the case where you just query by project name, right? we would have to fetch all models or latest model with the `project__` prefix

Reply:
Indeed. It is not possible to fetch the latest model from s3 / gcs as there is no way to sort files in descending order. (it's also not feasible to query for all models on every request to decide if we are still using the latest one).

Reply:
I've added a fix for this. Nevertheless, models that are added to the cloud storage after the server was started will not be considered.